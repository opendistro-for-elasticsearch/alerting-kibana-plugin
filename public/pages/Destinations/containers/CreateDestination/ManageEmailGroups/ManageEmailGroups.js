/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import _ from 'lodash';
import { Formik, FieldArray } from 'formik';
import {
  EuiButton,
  EuiButtonEmpty,
  EuiFlexGroup,
  EuiFlexItem,
  EuiHorizontalRule,
  EuiModal,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiModalBody,
  EuiModalFooter,
  EuiOverlayMask,
  EuiSpacer,
} from '@elastic/eui';
import PropTypes from 'prop-types';

import AddEmailGroupButton from '../../../components/createDestinations/AddEmailGroupButton/AddEmailGroupButton';
import EmailGroup from '../../../components/createDestinations/Email/EmailGroup';
import EmailGroupEmptyPrompt from '../../../components/createDestinations/EmailGroupEmptyPrompt';
import { emailGroupToFormik } from './utils/helpers';
import getEmailGroups from '../EmailRecipients/utils/helpers';
import { STATE } from '../../../components/createDestinations/Email/utils/constants';
import { ignoreEscape } from '../../../../../utils/helpers';

const createEmailGroupContext = (emailGroups) => ({
  ctx: {
    emailGroups,
  },
});

const getInitialValues = (emailGroups) =>
  _.isEmpty(emailGroups)
    ? { emailGroups: [] }
    : { emailGroups: emailGroups.map((emailGroup) => emailGroupToFormik(emailGroup)) };

const getEmailOptions = (emailGroups) => {
  if (_.isEmpty(emailGroups)) return [];

  // Get emails from email groups
  const nestedEmails = emailGroups.map((emailGroup) => emailGroup.emails);
  const emailOptions = [].concat(...nestedEmails);

  // Extract the email if it's already wrapped in a label (custom option)
  // so it can be filtered out
  const emails = emailOptions.map((email) => (_.isString(email) ? email : email.label));

  // Return a unique list of emails as options
  return [...new Set(emails)].map((email) => ({ label: email }));
};

export default class ManageEmailGroups extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      // initialValues is required in Formik v2, https://github.com/formium/formik/issues/1967
      initialValues: {},
      emailGroupsToDelete: [],
      loadingEmailGroups: true,
      failedEmailGroups: false,
    };
  }

  componentDidMount() {
    const { isEmailAllowed } = this.props;
    // Don't load initial values if Email is disallowed since the API will be blocked
    if (isEmailAllowed) {
      this.loadInitialValues();
    }
  }

  // Reload initial values when modal is no longer visible
  // or when email availability is updated
  componentDidUpdate(prevProps) {
    if (
      (prevProps.isVisible && !this.props.isVisible) ||
      prevProps.isEmailAllowed !== this.props.isEmailAllowed
    ) {
      this.loadInitialValues(this.props.emailGroups);
    }
  }

  loadInitialValues = async () => {
    const { httpClient } = this.props;
    this.setState({ loadingEmailGroups: true });

    const emailGroups = await getEmailGroups(httpClient);
    const initialValues = getInitialValues(emailGroups);

    this.setState({
      initialValues,
      emailGroupsToDelete: [],
      loadingEmailGroups: false,
    });
  };

  createEmailGroup = async (emailGroup) => {
    const { httpClient, notifications } = this.props;
    const body = {
      name: emailGroup.name,
      emails: emailGroup.emails.map((email) => ({ email: email.label })),
    };
    try {
      const response = await httpClient.post(`../api/alerting/destinations/email_groups`, {
        body: JSON.stringify(body),
      });
      if (!response.ok) {
        this.setState({ failedEmailGroups: true });
        notifications.toasts.addDanger({
          title: `Failed to create email group: ${emailGroup.name}`,
          text: `Reason: ${response.resp}`,
        });
      }
    } catch (err) {
      console.error('Unable to create email group', err);
      this.setState({ failedEmailGroups: true });
      notifications.toasts.addDanger({
        title: `Failed to create email group: ${emailGroup.name}`,
        text: `Reason: ${err}`,
      });
    }
  };

  updateEmailGroup = async (updatedEmailGroup) => {
    const { httpClient, notifications } = this.props;
    const { id, ifSeqNo, ifPrimaryTerm } = updatedEmailGroup;
    const body = {
      name: updatedEmailGroup.name,
      emails: updatedEmailGroup.emails.map((email) => ({ email: email.label })),
    };
    try {
      const response = await httpClient.put(`../api/alerting/destinations/email_groups/${id}`, {
        query: { ifSeqNo, ifPrimaryTerm },
        body: JSON.stringify(body),
      });
      if (!response.ok) {
        this.setState({ failedEmailGroups: true });
        notifications.toasts.addDanger({
          title: `Failed to update email group: ${updatedEmailGroup.name}`,
          text: `Reason: ${response.resp}`,
        });
      }
    } catch (err) {
      console.error('Unable to update email group', err);
      this.setState({ failedEmailGroups: true });
      notifications.toasts.addDanger({
        title: `Failed to update email group: ${updatedEmailGroup.name}`,
        text: `Reason: ${err}`,
      });
    }
  };

  deleteEmailGroup = async (emailGroup) => {
    const { httpClient, notifications } = this.props;
    const { id } = emailGroup;
    try {
      const response = await httpClient.delete(`../api/alerting/destinations/email_groups/${id}`);
      if (!response.ok) {
        this.setState({ failedEmailGroups: true });
        notifications.toasts.addDanger({
          title: `Failed to delete email group: ${emailGroup.name}`,
          text: `Reason: ${respons.resp}`,
        });
      }
    } catch (err) {
      console.err('Unable to delete email group', err);
      this.setState({ failedEmailGroups: true });
      notifications.toasts.addDanger({
        title: `Failed to delete email group: ${emailGroup.name}`,
        text: `Reason: ${err}`,
      });
    }
  };

  // TODO: Cleanup this function (currently making sequential API calls since each one has 'awaits' on it)
  processEmailGroups = async (values) => {
    const { emailGroupsToDelete } = this.state;
    const { emailGroups } = values;

    // Create or update email groups
    for (const emailGroup of emailGroups) {
      if (emailGroup.state === STATE.CREATED) {
        await this.createEmailGroup(emailGroup);
      } else if (emailGroup.state === STATE.UPDATED) {
        await this.updateEmailGroup(emailGroup);
      }
    }

    // Delete any removed email groups
    for (const emailGroup of emailGroupsToDelete) {
      await this.deleteEmailGroup(emailGroup);
    }

    // If there were no failures, show a success toast
    const { failedEmailGroups } = this.state;
    if (!failedEmailGroups) {
      this.props.notifications.toasts.addSuccess('Successfully saved email groups');
    }
    this.setState({ failedEmailGroups: false });
  };

  renderEmailGroups = ({ values, arrayHelpers }) => {
    const hasEmailGroups = !_.isEmpty(values.emailGroups);
    // TODO: Change this to getEmailOptions from stored emailGroups state
    const emailOptions = getEmailOptions(values.emailGroups);
    return hasEmailGroups ? (
      <div>
        {values.emailGroups.map((emailGroup, index) => (
          <div key={index}>
            <EmailGroup
              emailGroup={emailGroup}
              emailOptions={emailOptions}
              arrayHelpers={arrayHelpers}
              context={createEmailGroupContext(values.emailGroups)}
              index={index}
              onDelete={() => {
                if (emailGroup.id) {
                  this.setState((prevState) => ({
                    emailGroupsToDelete: [...prevState.emailGroupsToDelete, emailGroup],
                  }));
                }
                arrayHelpers.remove(index);
              }}
            />
            <EuiSpacer className="accordion-separator" />
          </div>
        ))}
        <div style={{ justifyContent: 'left' }}>
          <AddEmailGroupButton arrayHelpers={arrayHelpers} />
        </div>
      </div>
    ) : (
      <EmailGroupEmptyPrompt arrayHelpers={arrayHelpers} />
    );
  };

  hasEmailGroupsToProcess = (values) => {
    const { emailGroupsToDelete } = this.state;
    return !_.isEmpty(values.emailGroups) || !_.isEmpty(emailGroupsToDelete);
  };

  render() {
    const { isEmailAllowed, isVisible, onClickCancel, onClickSave } = this.props;
    const { initialValues, loadingEmailGroups } = this.state;
    return isVisible ? (
      <Formik
        initialValues={initialValues}
        onSubmit={(values, formikBag) => {
          this.processEmailGroups(values).then(() => onClickSave());
        }}
        validateOnChange={false}
        render={({ values, handleSubmit, isSubmitting }) => (
          <EuiOverlayMask>
            <EuiModal
              className="modal-manage-email"
              maxWidth={1000}
              onClose={ignoreEscape(onClickCancel)}
            >
              <EuiModalHeader>
                <EuiModalHeaderTitle>Manage email groups</EuiModalHeaderTitle>
              </EuiModalHeader>

              <EuiHorizontalRule margin="s" />

              <EuiModalBody>
                <FieldArray
                  name="emailGroups"
                  validateOnChange={true}
                  render={(arrayHelpers) =>
                    !isEmailAllowed || loadingEmailGroups ? (
                      <div style={{ display: 'flex', justifyContent: 'center' }}>
                        {isEmailAllowed ? 'Loading Email Groups...' : 'Email is disallowed'}
                      </div>
                    ) : (
                      this.renderEmailGroups({ values, arrayHelpers })
                    )
                  }
                />
              </EuiModalBody>

              {this.hasEmailGroupsToProcess(values) ? (
                <div>
                  <EuiHorizontalRule margin="s" />
                  <EuiModalFooter>
                    <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
                      <EuiFlexItem grow={false}>
                        <EuiButtonEmpty onClick={onClickCancel}>Cancel</EuiButtonEmpty>
                      </EuiFlexItem>
                      <EuiFlexItem grow={false}>
                        <EuiButton onClick={handleSubmit} isLoading={isSubmitting} fill>
                          Save
                        </EuiButton>
                      </EuiFlexItem>
                    </EuiFlexGroup>
                  </EuiModalFooter>
                </div>
              ) : null}
            </EuiModal>
          </EuiOverlayMask>
        )}
      />
    ) : null;
  }
}

ManageEmailGroups.propTypes = {
  httpClient: PropTypes.object.isRequired,
  isEmailAllowed: PropTypes.bool,
  isVisible: PropTypes.bool,
  onClickCancel: PropTypes.func,
  onClickSave: PropTypes.func,
  notifications: PropTypes.object.isRequired,
};

ManageEmailGroups.defaultProps = {
  isEmailAllowed: false,
  isVisible: false,
};
