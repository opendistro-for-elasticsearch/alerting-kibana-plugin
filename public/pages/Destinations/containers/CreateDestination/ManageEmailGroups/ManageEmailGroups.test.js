/*
 *   Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';

import ManageEmailGroups from './ManageEmailGroups';
import { httpClientMock } from '../../../../../../test/mocks';

const runAllPromises = () => new Promise(setImmediate);

const onClickCancel = jest.fn();
const onClickSave = jest.fn();

describe('ManageEmailGroups', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders', () => {
    const wrapper = mount(
      <ManageEmailGroups
        httpClient={httpClientMock}
        isVisible={false}
        onClickCancel={onClickCancel}
        onClickSave={onClickSave}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('renders when visible', () => {
    const wrapper = mount(
      <ManageEmailGroups
        httpClient={httpClientMock}
        isVisible={true}
        onClickCancel={onClickCancel}
        onClickSave={onClickSave}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('loadInitialValues', async () => {
    const mockEmailGroup = {
      id: 'id',
      name: 'test_group',
      emails: [{ email: 'test@email.com' }],
    };

    // Mock return in getEmailGroups function
    httpClientMock.get.mockResolvedValue({
      data: { ok: true, emailGroups: [mockEmailGroup] },
    });

    const wrapper = mount(
      <ManageEmailGroups
        httpClient={httpClientMock}
        isVisible={true}
        onClickCancel={onClickCancel}
        onClickSave={onClickSave}
      />
    );

    await runAllPromises();
    expect(wrapper.instance().state.initialValues.emailGroups.length).toBe(1);
    expect(wrapper.instance().state.initialValues.emailGroups[0].emails).toEqual([
      { label: 'test@email.com' },
    ]);
  });

  test('getEmailGroups logs resp.data.err when ok:false', async () => {
    const log = jest.spyOn(global.console, 'log');
    // Mock return in getEmailGroups function
    httpClientMock.get.mockResolvedValue({
      data: { ok: false, err: 'test' },
    });

    const wrapper = mount(
      <ManageEmailGroups
        httpClient={httpClientMock}
        isVisible={true}
        onClickCancel={onClickCancel}
        onClickSave={onClickSave}
      />
    );

    await runAllPromises();
    expect(log).toHaveBeenCalled();
    expect(log).toHaveBeenCalledWith('Unable to get email groups', 'test');
  });

  test('loads empty list of email groups when ok:false', async () => {
    // Mock return in getEmailGroups function
    httpClientMock.get.mockResolvedValue({
      data: { ok: false, err: 'test' },
    });

    const wrapper = mount(
      <ManageEmailGroups
        httpClient={httpClientMock}
        isVisible={true}
        onClickCancel={onClickCancel}
        onClickSave={onClickSave}
      />
    );

    await runAllPromises();
    expect(wrapper.instance().state.initialValues.emailGroups).toEqual([]);
  });
});
