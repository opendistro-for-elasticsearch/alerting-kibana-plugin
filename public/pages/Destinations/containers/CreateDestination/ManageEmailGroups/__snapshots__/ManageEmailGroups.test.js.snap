// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManageEmailGroups renders 1`] = `
<ManageEmailGroups
  httpClient={[MockFunction]}
  isEmailAllowed={true}
  isVisible={false}
  onClickCancel={[MockFunction]}
  onClickSave={[MockFunction]}
/>
`;

exports[`ManageEmailGroups renders when email is disallowed 1`] = `
<ManageEmailGroups
  httpClient={[MockFunction]}
  isEmailAllowed={false}
  isVisible={true}
  onClickCancel={[MockFunction]}
  onClickSave={[MockFunction]}
>
  <Formik
    enableReinitialize={false}
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={false}
  >
    <EuiOverlayMask>
      <Portal
        containerInfo={
          <div
            class="euiOverlayMask euiOverlayMask--aboveHeader"
          >
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="1"
            />
            <div
              data-focus-lock-disabled="false"
            >
              <div
                class="euiModal modal-manage-email"
                style="max-width: 1000px;"
                tabindex="0"
              >
                <button
                  aria-label="Closes this modal window"
                  class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                  type="button"
                >
                  <div>
                    EuiIconMock
                  </div>
                </button>
                <div
                  class="euiModal__flex"
                >
                  <div
                    class="euiModalHeader"
                  >
                    <div
                      class="euiModalHeader__title"
                    >
                      Manage email groups
                    </div>
                  </div>
                  <hr
                    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                  />
                  <div
                    class="euiModalBody"
                  >
                    <div
                      class="euiModalBody__overflow"
                    >
                      <div
                        style="display: flex; justify-content: center;"
                      >
                        Email is disallowed
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
          </div>
        }
      >
        <EuiModal
          className="modal-manage-email"
          maxWidth={1000}
          onClose={[Function]}
        >
          <EuiFocusTrap>
            <FocusLock
              as="div"
              autoFocus={true}
              disabled={false}
              lockProps={
                Object {
                  "style": undefined,
                }
              }
              noFocusGuards={false}
              persistentFocus={false}
              returnFocus={true}
            >
              <div
                data-focus-guard={true}
                key="guard-first"
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={0}
              />
              <div
                data-focus-guard={true}
                key="guard-nearest"
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={1}
              />
              <div
                data-focus-lock-disabled={false}
                onBlur={[Function]}
                onFocus={[Function]}
              >
                <SideEffect(FocusWatcher)
                  autoFocus={true}
                  disabled={false}
                  observed={
                    <div
                      data-focus-lock-disabled="false"
                    >
                      <div
                        class="euiModal modal-manage-email"
                        style="max-width: 1000px;"
                        tabindex="0"
                      >
                        <button
                          aria-label="Closes this modal window"
                          class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                          type="button"
                        >
                          <div>
                            EuiIconMock
                          </div>
                        </button>
                        <div
                          class="euiModal__flex"
                        >
                          <div
                            class="euiModalHeader"
                          >
                            <div
                              class="euiModalHeader__title"
                            >
                              Manage email groups
                            </div>
                          </div>
                          <hr
                            class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                          />
                          <div
                            class="euiModalBody"
                          >
                            <div
                              class="euiModalBody__overflow"
                            >
                              <div
                                style="display: flex; justify-content: center;"
                              >
                                Email is disallowed
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  }
                  onActivation={[Function]}
                  onDeactivation={[Function]}
                  persistentFocus={false}
                >
                  <FocusWatcher
                    autoFocus={true}
                    disabled={false}
                    observed={
                      <div
                        data-focus-lock-disabled="false"
                      >
                        <div
                          class="euiModal modal-manage-email"
                          style="max-width: 1000px;"
                          tabindex="0"
                        >
                          <button
                            aria-label="Closes this modal window"
                            class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                            type="button"
                          >
                            <div>
                              EuiIconMock
                            </div>
                          </button>
                          <div
                            class="euiModal__flex"
                          >
                            <div
                              class="euiModalHeader"
                            >
                              <div
                                class="euiModalHeader__title"
                              >
                                Manage email groups
                              </div>
                            </div>
                            <hr
                              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                            />
                            <div
                              class="euiModalBody"
                            >
                              <div
                                class="euiModalBody__overflow"
                              >
                                <div
                                  style="display: flex; justify-content: center;"
                                >
                                  Email is disallowed
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    }
                    onActivation={[Function]}
                    onDeactivation={[Function]}
                    persistentFocus={false}
                  />
                </SideEffect(FocusWatcher)>
                <div
                  className="euiModal modal-manage-email"
                  onKeyDown={[Function]}
                  style={
                    Object {
                      "maxWidth": "1000px",
                    }
                  }
                  tabIndex={0}
                >
                  <EuiI18n
                    default="Closes this modal window"
                    token="euiModal.closeModal"
                  >
                    <EuiButtonIcon
                      aria-label="Closes this modal window"
                      className="euiModal__closeIcon"
                      color="text"
                      iconType="cross"
                      onClick={[Function]}
                    >
                      <button
                        aria-label="Closes this modal window"
                        className="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                        onClick={[Function]}
                        type="button"
                      >
                        <EuiIcon
                          aria-hidden="true"
                          className="euiButtonIcon__icon"
                          size="m"
                          type="cross"
                        >
                          <div>
                            EuiIconMock
                          </div>
                        </EuiIcon>
                      </button>
                    </EuiButtonIcon>
                  </EuiI18n>
                  <div
                    className="euiModal__flex"
                  >
                    <EuiModalHeader>
                      <div
                        className="euiModalHeader"
                      >
                        <EuiModalHeaderTitle>
                          <div
                            className="euiModalHeader__title"
                          >
                            Manage email groups
                          </div>
                        </EuiModalHeaderTitle>
                      </div>
                    </EuiModalHeader>
                    <EuiHorizontalRule
                      margin="s"
                    >
                      <hr
                        className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                      />
                    </EuiHorizontalRule>
                    <EuiModalBody>
                      <div
                        className="euiModalBody"
                      >
                        <div
                          className="euiModalBody__overflow"
                        >
                          <C
                            name="emailGroups"
                            render={[Function]}
                            validateOnChange={true}
                          >
                            <FieldArrayInner
                              formik={
                                Object {
                                  "dirty": false,
                                  "errors": Object {},
                                  "handleBlur": [Function],
                                  "handleChange": [Function],
                                  "handleReset": [Function],
                                  "handleSubmit": [Function],
                                  "initialValues": Object {},
                                  "isSubmitting": false,
                                  "isValid": false,
                                  "isValidating": false,
                                  "registerField": [Function],
                                  "resetForm": [Function],
                                  "setError": [Function],
                                  "setErrors": [Function],
                                  "setFieldError": [Function],
                                  "setFieldTouched": [Function],
                                  "setFieldValue": [Function],
                                  "setFormikState": [Function],
                                  "setStatus": [Function],
                                  "setSubmitting": [Function],
                                  "setTouched": [Function],
                                  "setValues": [Function],
                                  "submitCount": 0,
                                  "submitForm": [Function],
                                  "touched": Object {},
                                  "unregisterField": [Function],
                                  "validate": undefined,
                                  "validateField": [Function],
                                  "validateForm": [Function],
                                  "validateOnBlur": true,
                                  "validateOnChange": false,
                                  "validationSchema": undefined,
                                  "values": Object {},
                                }
                              }
                              name="emailGroups"
                              render={[Function]}
                              validateOnChange={true}
                            >
                              <div
                                style={
                                  Object {
                                    "display": "flex",
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                Email is disallowed
                              </div>
                            </FieldArrayInner>
                          </C>
                        </div>
                      </div>
                    </EuiModalBody>
                  </div>
                </div>
              </div>
              <div
                data-focus-guard={true}
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={0}
              />
            </FocusLock>
          </EuiFocusTrap>
        </EuiModal>
      </Portal>
    </EuiOverlayMask>
  </Formik>
</ManageEmailGroups>
`;

exports[`ManageEmailGroups renders when visible 1`] = `
<ManageEmailGroups
  httpClient={[MockFunction]}
  isEmailAllowed={true}
  isVisible={true}
  onClickCancel={[MockFunction]}
  onClickSave={[MockFunction]}
>
  <Formik
    enableReinitialize={false}
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={false}
  >
    <EuiOverlayMask>
      <Portal
        containerInfo={
          <div
            class="euiOverlayMask euiOverlayMask--aboveHeader"
          >
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="1"
            />
            <div
              data-focus-lock-disabled="false"
            >
              <div
                class="euiModal modal-manage-email"
                style="max-width: 1000px;"
                tabindex="0"
              >
                <button
                  aria-label="Closes this modal window"
                  class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                  type="button"
                >
                  <div>
                    EuiIconMock
                  </div>
                </button>
                <div
                  class="euiModal__flex"
                >
                  <div
                    class="euiModalHeader"
                  >
                    <div
                      class="euiModalHeader__title"
                    >
                      Manage email groups
                    </div>
                  </div>
                  <hr
                    class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                  />
                  <div
                    class="euiModalBody"
                  >
                    <div
                      class="euiModalBody__overflow"
                    >
                      <div
                        style="display: flex; justify-content: center;"
                      >
                        Loading Email Groups...
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              data-focus-guard="true"
              style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
              tabindex="0"
            />
          </div>
        }
      >
        <EuiModal
          className="modal-manage-email"
          maxWidth={1000}
          onClose={[Function]}
        >
          <EuiFocusTrap>
            <FocusLock
              as="div"
              autoFocus={true}
              disabled={false}
              lockProps={
                Object {
                  "style": undefined,
                }
              }
              noFocusGuards={false}
              persistentFocus={false}
              returnFocus={true}
            >
              <div
                data-focus-guard={true}
                key="guard-first"
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={0}
              />
              <div
                data-focus-guard={true}
                key="guard-nearest"
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={1}
              />
              <div
                data-focus-lock-disabled={false}
                onBlur={[Function]}
                onFocus={[Function]}
              >
                <SideEffect(FocusWatcher)
                  autoFocus={true}
                  disabled={false}
                  observed={
                    <div
                      data-focus-lock-disabled="false"
                    >
                      <div
                        class="euiModal modal-manage-email"
                        style="max-width: 1000px;"
                        tabindex="0"
                      >
                        <button
                          aria-label="Closes this modal window"
                          class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                          type="button"
                        >
                          <div>
                            EuiIconMock
                          </div>
                        </button>
                        <div
                          class="euiModal__flex"
                        >
                          <div
                            class="euiModalHeader"
                          >
                            <div
                              class="euiModalHeader__title"
                            >
                              Manage email groups
                            </div>
                          </div>
                          <hr
                            class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                          />
                          <div
                            class="euiModalBody"
                          >
                            <div
                              class="euiModalBody__overflow"
                            >
                              <div
                                style="display: flex; justify-content: center;"
                              >
                                Loading Email Groups...
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  }
                  onActivation={[Function]}
                  onDeactivation={[Function]}
                  persistentFocus={false}
                >
                  <FocusWatcher
                    autoFocus={true}
                    disabled={false}
                    observed={
                      <div
                        data-focus-lock-disabled="false"
                      >
                        <div
                          class="euiModal modal-manage-email"
                          style="max-width: 1000px;"
                          tabindex="0"
                        >
                          <button
                            aria-label="Closes this modal window"
                            class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                            type="button"
                          >
                            <div>
                              EuiIconMock
                            </div>
                          </button>
                          <div
                            class="euiModal__flex"
                          >
                            <div
                              class="euiModalHeader"
                            >
                              <div
                                class="euiModalHeader__title"
                              >
                                Manage email groups
                              </div>
                            </div>
                            <hr
                              class="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                            />
                            <div
                              class="euiModalBody"
                            >
                              <div
                                class="euiModalBody__overflow"
                              >
                                <div
                                  style="display: flex; justify-content: center;"
                                >
                                  Loading Email Groups...
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    }
                    onActivation={[Function]}
                    onDeactivation={[Function]}
                    persistentFocus={false}
                  />
                </SideEffect(FocusWatcher)>
                <div
                  className="euiModal modal-manage-email"
                  onKeyDown={[Function]}
                  style={
                    Object {
                      "maxWidth": "1000px",
                    }
                  }
                  tabIndex={0}
                >
                  <EuiI18n
                    default="Closes this modal window"
                    token="euiModal.closeModal"
                  >
                    <EuiButtonIcon
                      aria-label="Closes this modal window"
                      className="euiModal__closeIcon"
                      color="text"
                      iconType="cross"
                      onClick={[Function]}
                    >
                      <button
                        aria-label="Closes this modal window"
                        className="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
                        onClick={[Function]}
                        type="button"
                      >
                        <EuiIcon
                          aria-hidden="true"
                          className="euiButtonIcon__icon"
                          size="m"
                          type="cross"
                        >
                          <div>
                            EuiIconMock
                          </div>
                        </EuiIcon>
                      </button>
                    </EuiButtonIcon>
                  </EuiI18n>
                  <div
                    className="euiModal__flex"
                  >
                    <EuiModalHeader>
                      <div
                        className="euiModalHeader"
                      >
                        <EuiModalHeaderTitle>
                          <div
                            className="euiModalHeader__title"
                          >
                            Manage email groups
                          </div>
                        </EuiModalHeaderTitle>
                      </div>
                    </EuiModalHeader>
                    <EuiHorizontalRule
                      margin="s"
                    >
                      <hr
                        className="euiHorizontalRule euiHorizontalRule--full euiHorizontalRule--marginSmall"
                      />
                    </EuiHorizontalRule>
                    <EuiModalBody>
                      <div
                        className="euiModalBody"
                      >
                        <div
                          className="euiModalBody__overflow"
                        >
                          <C
                            name="emailGroups"
                            render={[Function]}
                            validateOnChange={true}
                          >
                            <FieldArrayInner
                              formik={
                                Object {
                                  "dirty": false,
                                  "errors": Object {},
                                  "handleBlur": [Function],
                                  "handleChange": [Function],
                                  "handleReset": [Function],
                                  "handleSubmit": [Function],
                                  "initialValues": Object {},
                                  "isSubmitting": false,
                                  "isValid": false,
                                  "isValidating": false,
                                  "registerField": [Function],
                                  "resetForm": [Function],
                                  "setError": [Function],
                                  "setErrors": [Function],
                                  "setFieldError": [Function],
                                  "setFieldTouched": [Function],
                                  "setFieldValue": [Function],
                                  "setFormikState": [Function],
                                  "setStatus": [Function],
                                  "setSubmitting": [Function],
                                  "setTouched": [Function],
                                  "setValues": [Function],
                                  "submitCount": 0,
                                  "submitForm": [Function],
                                  "touched": Object {},
                                  "unregisterField": [Function],
                                  "validate": undefined,
                                  "validateField": [Function],
                                  "validateForm": [Function],
                                  "validateOnBlur": true,
                                  "validateOnChange": false,
                                  "validationSchema": undefined,
                                  "values": Object {},
                                }
                              }
                              name="emailGroups"
                              render={[Function]}
                              validateOnChange={true}
                            >
                              <div
                                style={
                                  Object {
                                    "display": "flex",
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                Loading Email Groups...
                              </div>
                            </FieldArrayInner>
                          </C>
                        </div>
                      </div>
                    </EuiModalBody>
                  </div>
                </div>
              </div>
              <div
                data-focus-guard={true}
                style={
                  Object {
                    "height": "0px",
                    "left": "1px",
                    "overflow": "hidden",
                    "padding": 0,
                    "position": "fixed",
                    "top": "1px",
                    "width": "1px",
                  }
                }
                tabIndex={0}
              />
            </FocusLock>
          </EuiFocusTrap>
        </EuiModal>
      </Portal>
    </EuiOverlayMask>
  </Formik>
</ManageEmailGroups>
`;
